Implement BuildConfig reaper

mnagy opened the pull request:
The reaper will delete all builds associated with the BC.

This is another shot at #5080. That one was closed, but after discussion at #5208, it became necessary to return to the previous attempt. Differences from #5080 are that we now handle deprecated labels, I've added unit tests and also extended tests from #5208.

mnagy commented:
[test][extended(builds)]

kargakis commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	// Collect builds related to the config.
+	buildList, err := reaper.oc.Builds(namespace).List(util.ConfigSelector(name), nil)
```
Why two lists instead of one and filter twice?

kargakis commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,66 @@
+		return "", kerrors.NewNotFound("BuildConfig", name)
+	}
+	for _, build := range builds {
+		if err = reaper.oc.Builds(namespace).Delete(build.Name); err != nil {
```
We probably need better error handling here. Shouldn't print anything if the build wasn't found.

soltysh replied:
I'd suggest solution similar to what we have [upstream](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/stop.go#L259-L267), iow. collect errors in return NewAggregate if there are errors.

soltysh replied:
There's one significant problem with this approach. If there's currently running build pod and you remove the build object, the pod gets orphaned. I'm aware there's a logic to handle that, but that's not how you should handle this. Upon stop you should remove everything that belongs to this build. See how deployments are invoking the replication controller reaper in the end, which handles that part. Since build controller manually start pods you need to manually invoke pod reaper as well. And don't forget to create a test covering such situation!

mnagy replied:
@bparees any thoughts on @soltysh's comments? Options we have as I see it:
 1. Only delete builds, leave pods to the controller,
 2. Delete builds and any left-over build pods,
 3. Add a build reaper that would delete left-over build pods and use that reaper in BC reaper.

mnagy replied:
I think I lean on the option to leave it as is and let the controller handle pods. It'll avoid duplicating the deletion logic and we can rework the whole thing when server cascading is worked out.

soltysh commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,195 @@
+				t.Errorf("%s: unexpected action: %s, expected %s", test.testName, actualAction, test.expected[j])
+			}
+		}
+		if out != test.output {
```
Drop this test, we've changed the Reaper interface upstream already, currently it returns only error. This will cause me problems upon next rebase ;)

soltysh commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,195 @@
+		expected  []ktestclient.Action
+		output    string
+		err       bool
+	}{
```
Instead of embedding a testName into struct do `map[string]struct{}`.
Namespace is not changing, hardcode it. 
Drop output, as stated further.

rhcarvalho replied:
> Instead of embedding a testName into struct do `map[string]struct{}`.

I find it a bit less convenient to use this type of more evolved table-driven test with a map.
Maps have no guaranteed order; don't allow us to print a consistent index to the test case; make it impossible to, while debugging, run only a fraction of the tests by slicing the test cases, ...

And I haven't seen people using maps for that purpose in other projects in the Go community.

On the plus side, maps wouldn't allow you to accidentally give the same name to two or more test cases.

There are no occurrences of `map[string]struct` in Go's standard library, while there are some table-driven tests using a name field to describe the test.

- https://github.com/golang/go/blob/86f1944e86ae3e07651dcbfafee37f1e3b0a917d/src/cmd/api/goapi_test.go#L82
- https://github.com/golang/go/blob/86f1944e86ae3e07651dcbfafee37f1e3b0a917d/src/crypto/cipher/cbc_aes_test.go#L29

@soltysh why are you advocating using maps?

kargakis replied:
+1 for table-driven tests using slices instead of maps. Maps are nice but we shouldn't be so smart in tests for the reasons @rhcarvalho mentioned, (even though most of the time we don't need any kind of indexing, it's good to have it).

liggitt replied:
-1 for slices... finding a test by name is much easier

rhcarvalho replied:
@liggitt the question is not naming a test case or not, but:

```go
tests := map[string]struct {
	a, b, c int
}{
	"test 1": {
		a: 1,
		b: 2,
		c: 3,
	},
	"test 2": {
		a: 4,
		b: 5,
		c: 6,
	},
}
```

vs.

```go
tests := []struct {
	name    string
	a, b, c int
}{
	{
		name: "test 1",
		a:    1,
		b:    2,
		c:    3,
	},
	{
		name: "test 2",
		a:    4,
		b:    5,
		c:    6,
	},
}
```

rhcarvalho replied:
I don't want us to go into bike shedding. Just asked because since @soltysh was advocating one over the other, I thought there might be good arguments.

liggitt replied:
I understand, but you want unique names for tests, a map guarantees that. I also find the name with the test data indented more readable

liggitt replied:
I also like not having a guaranteed order of execution... makes you write side-effect free isolated tests or risk flakiness

soltysh commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,195 @@
+	"github.com/openshift/origin/pkg/client/testclient"
+)
+
+func mkbuildconfig(version int) *buildapi.BuildConfig {
```
I don't like this naming scheme. Please stick with what we have mostly, iow. `makeBuildConfig`. Yes, I'm aware the deployment reaper looks like that ;)

soltysh commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,195 @@
+}
+
+func TestStop(t *testing.T) {
+	notfound := func() runtime.Object {
```
Again naming scheme, we follow camelCase, so notFound, please :smiley_cat:

kargakis commented:
Some thoughts:
* @soltysh after talking with @mnagy and since the controllers handle pod deletion in case of a missing build then I think we shouldn't create a build reaper.
* the flow of deleting a bc can borrow from the one from jobs - both resources can have many childs so it would be wiser to delete them in the end of the process (see https://github.com/kubernetes/kubernetes/pull/18299#issuecomment-162659692). The major thing here is to adopt a way to say to the build controllers that "this bc is marked for deletion so don't attempt to create any new build for it" (probably an annotation? This is achieved in jobs much simpler by setting parallelism to zero which means "don't run any pods for this job"). This would affect probably all build cli commands. Then the reaper flow would be: 1) mark this bc for deletion, 2) delete all builds, 3) delete bc.

mnagy commented:
Addressed @soltysh's test comments and expanded the loop in the reaper as @kargakis wanted, which makes it a bit more verbose but also cleaner I think. I left the overall logic for deleting as it was.

Are there any more comments on that? @bparees?

mfojtik commented on pkg/build/controller/strategy/custom.go:
```diff
@@ -22,6 +22,7 @@ type CustomBuildStrategy struct {
 
 // CreateBuildPod creates the pod to be used for the Custom build
 func (bs *CustomBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod, error) {
+	build.Status.Config.APIVersion
```
what is this doing?

mfojtik commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,74 @@
+	//
+	// We delete the config first to ensure no further builds will be created.
+	err := reaper.oc.BuildConfigs(namespace).Delete(name)
+	notFound := kerrors.IsNotFound(err)
```
why you need a special variable to assign this error to?

mnagy replied:
So it can be unset later. The point is that we only return not found error if no BC nor Builds are found. If BC is not found but some Builds are, that means we're just finishing up a previous deletion, so it's not an error.

kargakis replied:
@deads2k and @smarterclayton convinced me this is not right (for jobs) and since there are no races with the BC controller I would vote for moving the bc deletion in the end. Hopefully, making the BC controller aware of a bc that is about to be deleted won't be difficult?

mnagy replied:
You mean setting the annotation?

kargakis replied:
Yes. I want to read opinions from @bparees and @mfojtik on this though

deads2k replied:
Being able to cancel my delete and start the delete again and still work has value.

smarterclayton replied:
We should always reap the main object at the end.

On Wed, Dec 16, 2015 at 11:55 AM, Michail Kargakis <notifications@github.com
> wrote:

> In pkg/build/reaper/reaper.go
> <https://github.com/openshift/origin/pull/6185#discussion_r47801883>:
>
> > +// BuildConfigReaper implements the Reaper interface for buildConfigs
> > +type BuildConfigReaper struct {
> > +	oc                    client.Interface
> > +	pollInterval, timeout time.Duration
> > +}
> > +
> > +// Stop deletes the build configuration and all of the associated builds.
> > +func (reaper *BuildConfigReaper) Stop(namespace, name string, timeout time.Duration, gracePeriod *kapi.DeleteOptions) (string, error) {
> > +	// If the config is already deleted, it may still have associated
> > +	// builds which didn't get cleaned up during prior calls to Stop. If
> > +	// the config can't be found, still make an attempt to clean up the
> > +	// builds.
> > +	//
> > +	// We delete the config first to ensure no further builds will be created.
> > +	err := reaper.oc.BuildConfigs(namespace).Delete(name)
> > +	notFound := kerrors.IsNotFound(err)
>
> @deads2k <https://github.com/deads2k> and @smarterclayton
> <https://github.com/smarterclayton> convinced me this is not right (for
> jobs) and since there are no races with the BC controller I would vote for
> moving the bc deletion in the end. Hopefully, making the BC controller
> aware of a bc that is about to be deleted won't be difficult?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6185/files#r47801883>.
>

bparees replied:
@kargakis can you summarize the argument for why it's better to have complicated quiesce logic so we can delete the BC at the end, instead of just deleting the BC immediately?  Is the concern about leaving things in a half-cleaned up state which the user might not notice if the BC Is gone?

I guess i don't have a strong aversion to adding a "quiesced/deletepending" type annotation, but it does add a whole new range of error messages we need to return to people who try to instantiate one.

kargakis replied:
@bparees see https://github.com/openshift/origin/pull/6185#issuecomment-165123571 and the discussion in https://github.com/kubernetes/kubernetes/pull/18299 for context. I am not oversold to it but since we are stuck currently with cascading in the client, I think it's slightly better than the opposite (what the DC reaper is doing).

bparees replied:
ok i've read through it, @mnagy i'm afraid you're going to need to implement buildconfig quiescing and delete the BC at the end.

mfojtik commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,74 @@
+	errList := []error{}
+	for _, build := range builds.Items {
+		notFound = false
+		if err = reaper.oc.Builds(namespace).Delete(build.Name); err != nil && !kerrors.IsNotFound(err) {
```
is this skipping the not found errors intentionally? should be at least log it?

kargakis replied:
Yes, we should log this at least under a v4 level though

mnagy replied:
Yes, it is intentional. Logging at V4 is fine IMHO.

mfojtik commented on test/extended/builds/remove_buildconfig.go:
```diff
@@ -0,0 +1,68 @@
+				// Note that it's not important to check for success here. We
+				// only care about builds being deleted after BC is deleted and
+				// we don't really care about their status prior to that.
+				exutil.WaitForABuild(oc.REST().Builds(oc.Namespace()), builds[i], exutil.CheckBuildSuccessFn, exutil.CheckBuildFailedFn)
```
you can do start-build --wait  and get rid of this (start-build --wait waits till the build complete (or fail))

mnagy replied:
But I specifically don't want to wait for two other builds (hence len(builds)/2 in the range up). The idea here is that you don't care about states of builds you are removing.

rhcarvalho replied:
Hmm what's special about len(builds)/2?

rhcarvalho replied:
Why do we need to wait for a build to start? In that case, wouldn't it be better to wait for only 1 build?
And instead of waiting for a specific build of the array, we should simply wait for "any one" build to finish, maybe by looking at `oc get builds` or something else?

Waiting for one particular build might make us unintentionally wait for all builds, depending on how builds are scheduled.

mnagy replied:
Hmm, I guess you're right that waiting for one build should be enough. But I think waiting for the first one should be sufficient. I know it might make us wait for all builds, but I think this would be arbitrary and at worst produce a flaky behaviour.

mnagy replied:
I meant first in the array, not first one that finishes.

rhcarvalho commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,174 @@
+	}
+}
+
+func mkbuildlist(version int) *buildapi.BuildList {
```
This one needs to be renamed as @soltysh pointed out earlier.

mnagy replied:
oops, missed that one :)

rhcarvalho commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,174 @@
+
+	for testName, test := range tests {
+		reaper := &BuildConfigReaper{oc: test.oc, pollInterval: time.Millisecond, timeout: time.Millisecond}
+		_, err := reaper.Stop("default", test.name, 1*time.Second, nil)
```
Hmm you bought the idea to hard code the namespace, "default", but now we have all tests with the same `name: "config"`...
Can we tell that all tests should likely run under the same namespace and name? IOW, does it matter for what is being tested here to be able to use a different namespace/name?

kargakis replied:
I don't think we really care about the namespace/name of the object being tested as long as the test case has a name.

mnagy replied:
Ah, missed that. Will remove name as well.

bparees commented on pkg/build/util/util.go:
```diff
@@ -70,3 +71,15 @@ func IsBuildComplete(build *buildapi.Build) bool {
+// ConfigSelector returns a label Selector which can be used to find all
+// builds for a BuildConfig.
+func ConfigSelector(name string) labels.Selector {
+	return labels.Set{buildapi.BuildConfigLabel: name}.AsSelector()
```
we don't have this somewhere else already?

mnagy replied:
I don't believe we have. You might have seen it in my other PR's that didn't get merged.

mfojtik replied:
need better name, "BuildConfigSelector" as this is related to BuildConfig

mnagy commented:
I have added a new annotation that is used when deleting the BC and deleting it at the end. Tests updated as well. PTAL

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -199,6 +199,10 @@ func (g *BuildGenerator) Instantiate(ctx kapi.Context, request *buildapi.BuildRe
 
+	if _, ok := bc.Annotations[buildapi.BuildConfigDeletionAnnotation]; ok {
+		return nil, fmt.Errorf("can't instantiate from BuildConfig %s/%s: BuildConfig is being deleted", bc.Namespace, bc.Name)
+	}
```
Is this error going to get retried by the retry controller?  we want to make sure this is treated as a FatalError, otherwise we're going to pointlessly retry it.

see:
https://github.com/openshift/origin/blob/master/pkg/build/controller/factory/factory.go#L320-L331
https://github.com/openshift/origin/blob/master/pkg/build/controller/factory/factory.go#L159-L178

mnagy replied:
Nice catch.

kargakis commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,107 @@
+	noBuildFound := true
+
+	// Add deletion pending annotation to the build config
+	for {
```
You may want to use a more sophisticated polling mechanism than a simple for (eg. [wait.Poll](https://github.com/kubernetes/kubernetes/blob/c8f1019c1064773727f1adccee275dfd3e1d00e8/pkg/util/wait/wait.go#L48), which adds a timeout).

liggitt replied:
Or PollImmediate

mnagy replied:
Why? It will only try again if there is a conflict. Which I assume can't happen indefinitely many times.

mfojtik replied:
+1 to PollImmediate

mnagy replied:
What is the benefit gained? I think that for is more idiomatic here. We do not require polling nor timeout here. Using a polling function would only make the code more complicated and harder to follow.

deads2k replied:
This is `RetryOnConflict` (https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/conditions.go#L63) which will eliminate one check, prevent the hot looping, and make the conditions a little easier to follow.

mnagy replied:
Cool, didn't know about that one.

kargakis commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	// Collect builds related to the config.
+	builds, err := reaper.oc.Builds(namespace).List(buildutil.ConfigSelector(name), nil)
```
Since we know that the relation here is based on the name, do we really need this list?
Can't you try to delete all builds by having just the bc?

liggitt replied:
Don't you have to list builds in order to iterate and delete?

kargakis replied:
Well, technically it is not needed but thinking about it a bit more, if any pruning has happened before then we would try to delete a bunch of non-existent builds so.... better list:)

mnagy commented:
@bparees addressed the fatal error issue, PTAL

mnagy commented:
The travis fail unrelated to this PR.

bparees commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,107 @@
+		if err != nil {
+			errList = append(errList, err)
+		}
+	}
```
we do this even if there were errors deleting some of the builds?

bparees commented:
one final question, otherwise lgtm.

mnagy commented:
Good point, we shouldn't delete the BC if deletion of a build failed. I've restructured the reaper to only delete it if nothing fails. However, this will leave us with BC with deletion annotation and the user will need to finish deletion manually.

I'm thinking now that this should be handled by a big warning to the user, or something to that effect.

mfojtik commented on pkg/build/api/types.go:
```diff
@@ -409,6 +409,8 @@ const (
 	// BuildConfigLabelDeprecated was used as BuildConfigLabel before adding namespaces.
 	// We keep it for backward compatibility.
 	BuildConfigLabelDeprecated = "buildconfig"
+	// BuildConfigDeletionAnnotation is an annotation that identifies a BuildConfig as being deleted
```
marks a BuildConfig that is being deleted

mfojtik commented on pkg/build/controller/config_controller.go:
```diff
@@ -10,8 +10,18 @@ import (
 
+// ConfigControllerFatalError represents a fatal error while generating a build
+type ConfigControllerFatalError struct {
+	Reason string
```
godoc the reason as well

mfojtik commented on pkg/build/controller/config_controller.go:
```diff
@@ -10,8 +10,18 @@ import (
+	Reason string
+}
+
+func (e ConfigControllerFatalError) Error() string {
```
godoc this as it is exported, also explain in godoc what does the FatalError means

mfojtik commented on pkg/build/controller/config_controller.go:
```diff
@@ -50,6 +60,8 @@ func (c *BuildConfigController) HandleBuildConfig(bc *buildapi.BuildConfig) erro
 		if kerrors.IsConflict(err) {
 			instantiateErr = fmt.Errorf("unable to instantiate Build for BuildConfig %s/%s due to a conflicting update: %v", bc.Namespace, bc.Name, err)
 			util.HandleError(instantiateErr)
+		} else if _, ok := err.(buildgenerator.GeneratorFatalError); ok {
```
maybe we need something like `IsFatalError` ?

mfojtik commented on pkg/build/controller/factory/factory.go:
```diff
@@ -322,7 +322,10 @@ func (factory *BuildConfigControllerFactory) Create() controller.RunnableControl
 			cache.MetaNamespaceKeyFunc,
-			retryFunc("BuildConfig", nil),
+			retryFunc("BuildConfig", func(err error) bool {
+				_, isFatal := err.(buildcontroller.ConfigControllerFatalError)
```
and you can reuse that IsFatal here :-)

mfojtik commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,107 @@
+	if err != nil {
+		return "", err
+	}
+	errList := []error{}
```
i thought we already have something like ErrorList in k8s

mfojtik commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,107 @@
+	errList := []error{}
+	for _, build := range builds.Items {
+		noBuildFound = false
+		if err = reaper.oc.Builds(namespace).Delete(build.Name); err != nil {
```
err :=

mfojtik commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,107 @@
+	}
+	for _, build := range builds.Items {
+		noBuildFound = false
+		if err = reaper.oc.Builds(namespace).Delete(build.Name); err != nil {
```
if err :=

mfojtik commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,107 @@
+	// Finally we can delete the BuildConfig
+	if !noBcFound {
+		err = reaper.oc.BuildConfigs(namespace).Delete(name)
+		if err != nil {
```
this is a perfect case when you should use `if err:= ; err != nil`...

mfojtik commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,194 @@
+func makeBuildConfig(version int, deleting bool) *buildapi.BuildConfig {
+	ret := &buildapi.BuildConfig{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:        "config",
```
make a better name?

mnagy replied:
This is fine for unit test. It's borrowed from the deploy reaper tests.

mfojtik commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,194 @@
+	"github.com/openshift/origin/pkg/client/testclient"
+)
+
+func makeBuildConfig(version int, deleting bool) *buildapi.BuildConfig {
```
what does "deleting" do?

bparees replied:
makes a buildconfig with the "being deleted" annotation set...

mfojtik replied:
`setDeletingAnnotation bool` then?

mfojtik replied:
but since this is a test, you can ignore me :-)

mfojtik commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,194 @@
+	return false
+}
+
+func TestStop(t *testing.T) {
```
give it better name

mnagy replied:
I think it's a good name, it tests a Stop function. Also, borrowed from deploy reaper test.

mfojtik commented on pkg/build/reaper/reaper_test.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func TestStop(t *testing.T) {
+	notFound := func() runtime.Object {
```
why this function has to be stored when you use it just once?

mnagy replied:
Borrowed from deploy. I used it at two places before, not sure when I removed the one. But I still like it this way. It does not need to clutter up test tables and can be easily reused if new tests are added. I don't see any harm in it in any case.

mfojtik commented:
some coding nits, please address them before merge, otherwise LGTM

liggitt commented:
Reapers need to be built to be reentrant. If the reap fails, it should be
able to be retried successfully.

On Jan 8, 2016, at 8:35 AM, Martin Nagy <notifications@github.com> wrote:

Good point, we shouldn't delete the BC if deletion of a build failed. I've
restructured the reaper to only delete it if nothing fails. However, this
will leave us with BC with deletion annotation and the user will need to
finish deletion manually.

I'm thinking now that this should be handled by a big warning to the user,
or something to that effect.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/6185#issuecomment-170005306>.

bparees commented:
I don't think the deletion annotation existing prevents retrying it.  The only question is should the buildconfig be left in a "Deletion pending" state (no new builds can be created) or reverted to a normal state until the user retries the deletion.

i'm inclined to say we should leave it in deletion pending.  that leaves it as a stale artifact that can't be used for anything, but may need admin action to clean up, which is better than leaving it in a state where new builds might get created.

So what i'm suggesting is:
1) don't delete the BC if you can't delete all the builds
2) leave the annotation in place
3) allow this to be called again on a BC that has the annotation (I think that's already the case?)
4) Warn the user when we didn't successfully delete everything.

The only concern is what happens if two people try to delete the BC simultaneously.  Since it's reentrant, we'd allow it, but they're likely to get errors during the build delete process.  I guess that's ok, they can just call delete BC a third time.

bparees commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,107 @@
+			return "", err
+		}
+		err = util.AddObjectAnnotations(bc, map[string]string{
+			buildapi.BuildConfigDeletionAnnotation: "",
```
since we're going to allow the BC to stick around if something went wrong, but with the annotation in place, this logic needs to check if the annotation already exists and not add if it so, so we don't end up with it duplicated.

mnagy commented:
@bparees I agree with you. Clearing up the deletion pending annotation could leave things in a weird state (e.g. some random builds deleted). Plus the clearing up could fail anyway, so we would need to be prepared for that situation anyway.

Your points: 1 through 3 are true  now, 4 is missing and I will add it.

bparees commented:
the current code doesn't implement point 1 as i'm reading it, it always deletes the BC.

bparees commented:
sorry nm, i missed the new return block.

mnagy commented:
Done some refactoring, addressed comments (except the nits for tests). Added a warning for the user that makes it clear that the BC was not deleted.

bparees commented:
> Added a warning for the user that makes it clear that the BC was not deleted.

that warning is going to show up in the server logs, right?  No user is going to see that.

What is the user experience when they run oc delete bc/foo and it does not delete the bc?

mnagy commented:
@bparees this is client-side code, so user should see the warning

bparees commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,113 @@
+
+	// Aggregate all errors
+	if len(errList) > 0 {
+		glog.Warningf("Some errors occurred, BuildConfig %s/%s was not deleted", namespace, name)
```
This warning needs to give the user some clue what went wrong and what they should do about it.

mnagy replied:
Well, more errors will be in the log. Should I list the error list as well? We could also ask the user to try again.

bparees replied:
i'd at least tell them we failed to delete some builds and maybe list the builds we failed to delete.

"Failed to delete all builds associated with the buildconfig, so the buildconfig was not deleted.  Try deleting the builds manually.  Builds that were not deleted: <list builds>"

mnagy replied:
Hmm. Not sure I would want to list all the builds. How about telling the user it can try to re-run the command with additional verbosity? The builds will either get deleted then, or fail with the same error, so the user will see the errors.

bparees replied:
sure that's a good idea.

mnagy commented:
I realized we should warn the user about failed builds right away anyway, so I changed previous V(4) info messages into warnings and added a hint about removing the builds manually at the end.

bparees commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,113 @@
+
+	// Aggregate all errors
+	if len(errList) > 0 {
+		glog.Warningf("Failed to delete all Builds associated with BuildConfig %s/%s. You can try re-running the command or removing them manually", namespace, name)
```
"BuildConfig %s/%s will not be deleted because not all associated builds could be deleted.  You can try....."

(point being you need to tell them the buildconfig was not deleted)

mnagy commented:
ptal

bparees commented:
lgtm, couple nits from @mfojtik left to be addressed in the tests.

mnagy commented:
@bparees I didn't addressed those deliberately. I've added comments with explanations (except for the one case where @mfojtik explicitly said it can be ignored).

mfojtik commented:
LGTM @mnagy is this mergeable now?

mnagy commented:
@mfojtik yes, but I think this might need a @openshift/api-review ack for added annotation.

rhcarvalho commented on test/extended/builds/remove_buildconfig.go:
```diff
@@ -0,0 +1,68 @@
+	})
+
+	g.Describe("oc delete buildconfig", func() {
+		g.It("should start a build and wait for the build to complete", func() {
```
"oc delete buildconfig should start a build and wait for the build to complete" ?!

I thought it should delete a buildconfig.

mnagy replied:
oops :) copy&paste error

deads2k commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,113 @@
+	for {
+		bc, err := reaper.oc.BuildConfigs(namespace).Get(name)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
```
This can be promoted outside of the `if err != nil` to keep the code flatter.

deads2k commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,113 @@
+	errList := []error{}
+	for _, build := range builds.Items {
+		noBuildFound = false
+		if err := reaper.oc.Builds(namespace).Delete(build.Name); err != nil {
```
Question: Do we have a controller that cleans up orphaned build pods server side?  If not, this should probably ripple all the way down.   I'd be fine with a followup.

mnagy replied:
Yes, here: https://github.com/openshift/origin/blob/master/pkg/build/controller/controller.go#L347

deads2k commented on pkg/build/api/types.go:
```diff
@@ -411,6 +411,8 @@ const (
 	// We keep it for backward compatibility.
 	BuildConfigLabelDeprecated = "buildconfig"
+	// BuildConfigDeletionAnnotation is an annotation that marks a BuildConfig that is being deleted
+	BuildConfigDeletionAnnotation = "openshift.io/build-config.deletion-pending"
```
The usage on this field looks more like "inert" than deleting, right?  This stops you from creating builds based on this bc, but no controller actually does deletion or reaping server-side based on this annotation.

We've talked about making deployments inert to keep them from doing work, this looks parallel to that.

Also, I'd like to set it to the value "true" (detection being case insensitive).  Existence is definitive, but not as explicit for users.

kargakis replied:
> We've talked about making deployments inert to keep them from doing work, this looks parallel to that.

@ironcladlou @deads2k does this mean that the dc controller should ignore altogether a deployment (that is explicitly marked as inert via a spec field) or the lack of a config change trigger is enough for the deployment to be considered inert?

deads2k replied:
> @ironcladlou @deads2k does this mean that the dc controller should ignore altogether a deployment (that is explicitly marked as inert via a spec field) or the lack of a config change trigger is enough for the deployment to be considered inert?

I meant upstream.  I haven't heard that we'll be pulling it downstream, though given the upstream lack of image change triggers, I don' t see our DCs going away for some time.

I chose the example for vocabulary.

ironcladlou replied:
>@ironcladlou @deads2k does this mean that the dc controller should ignore altogether a deployment (that is explicitly marked as inert via a spec field) or the lack of a config change trigger is enough for the deployment to be considered inert?

You'd have to delete the configChange trigger and set any image triggers to "automatic=false".

mnagy replied:
Like this?
```
BuildConfigInertAnnotation = "openshift.io/build-config.inert"
```

kargakis replied:
> You'd have to delete the configChange trigger and set any image triggers to "automatic=false".

Right, I thought @deads2k was talking about DCs.

Upstream PR for inert deployments is https://github.com/kubernetes/kubernetes/pull/18901

kargakis replied:
If it makes sense to have inert buildConfigs then I would vote for a field in the spec vs an annotation.

deads2k replied:
> If it makes sense to have inert buildConfigs then I would vote for a field in the spec vs an annotation.

I'd rather use an annotation until its baked upstream so we can be sure of consistency.

mnagy commented:
Updated. Using a "paused" annotation with a "true" value. Removed waiting for builds in extended tests as discussed with @rhcarvalho off-github. This should make things a lot faster. Also restructured the conflict-avoiding loop as suggested by @deads2k.

PTAL

bparees commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,111 @@
+	noBuildFound := true
+
+	// Add deletion pending annotation to the build config
+	unversioned.RetryOnConflict(unversioned.DefaultRetry, func() error {
```
you should still record the error this returns, if any, so you can report it (in the case that it's not a conflict error)

bparees commented:
at this point i'd like to hold merging this until 3.1.1 gets out the door.  but pending the error handling comment i just made, lgtm.

bparees commented on pkg/build/reaper/reaper.go:
```diff
@@ -0,0 +1,114 @@
+	builds, err = reaper.oc.Builds(namespace).List(buildutil.BuildConfigSelectorDeprecated(name), nil)
+	if err != nil {
+		return "", err
+	}
```
if you error out here, you're not going to print the warning that the BuildConfig isn't being deleted.

I think you need to audit all the return calls in this function to make sure the warning is always getting printed.

mnagy replied:
I fixed it by adding a deferred function. PTAL

bparees replied:
oooh, i like it!

mfojtik replied:
s/Note/TODO:/ :D

mnagy replied:
@mfojtik nit-picker :) Done ;)

mnagy commented:
@openshift/ui-review fyi, this PR is adding an annotation that will make a BC inert, disallowing new builds to be instantiated from the BC. This should only be set while the deletion is happening, but might stay there if an error occurs.

jhadvig commented:
@mnagy thanks for heads up. Once the PR is merged will adjust the logic in UI.

mnagy commented:
@bparees can you remove the DO_NOT_MERGE tag now?

bparees commented:
@mnagy will do but i'm still waiting for the official merge queue open email (unlike everyone else, apparently)

bparees commented:
[merge]

bparees commented:
imagestream import flake.

bparees commented:
congrats @mnagy :wine_glass:
